### ISSUE TITLE
csum exists for 31516807168-31521120256 but there is no extent record

### ISSUE THREAD

**Issue by Vochatrak-az-ezm** [kdave/btrfs-progs#430]
Errors were found on the btrfs partition:

```
# btrfs check --force --readonly /dev/sdd2
Opening filesystem to check...
Checking filesystem on /dev/sdd2
UUID: 0e0e2f5d-a93f-4633-ba12-15a06763d641
[1/7] checking root items
[2/7] checking extents
[3/7] checking free space cache
[4/7] checking fs roots
[5/7] checking only csums items (without verifying data)
there are no extents for csum range 31516819456-31516872704
csum exists for 31516807168-31521120256 but there is no extent record
ERROR: errors found in csum tree
[6/7] checking root refs
[7/7] checking quota groups skipped (not enabled on this FS)
found 685500407808 bytes used, error(s) found
total csum bytes: 587788724
total tree bytes: 753860608
total fs tree bytes: 34062336
total extent tree bytes: 27377664
btree space waste bytes: 109059044
file data blocks allocated: 688818327552
 referenced 690986815488
```

I tried to reset the cache, it didn't help.

scrub said that everything is fine.

Everything is written/read normally on the section.

There are actually two questions:

1). How to treat this without losing data?

2). Can I leave it as it is (it seems to work the same)? I won't get any problems?

ArcoLinux 
Kernel Version 5.15.8

**Comment by adam900710** [comment]
This is strange, as when we delete one data extent, we also delete its csum at the same time, inside the same transaction.
Thus it should be some extra path causing the problem.

For your questions:

1. btrfs check --init-csum-tree
   Since you have already run scrub, your existing data is all correct, thus we can rebuild csum tree using the known good data.

2. Properly no
  The problem here is, when new data extent is allocated for that range, and it has csum, it will cause -EEXIST error and cause a failure.

**Comment by adam900710** [comment]
Could you provide some history of the fs?

Like the oldest kernel it has been run on, or does it experience some btrfs check --repair?

**Comment by Vochatrak-az-ezm** [comment]
Yes, there was no special story. Used in normal mode.

fstab looked like this:

```
UUID=0e0e2f5d-a93f-4633-ba12-15a06763d641 /mnt/Disk_D/App btrfs subvol=@App,defaults,max_inline=256,noatime,space_cache=v2,autodefrag,compress=zstd:15 0 0
#
UUID=0e0e2f5d-a93f-4633-ba12-15a06763d641 /mnt/Disk_D/AppImages btrfs subvol=@AppImages,defaults,max_inline=256,noatime,space_cache=v2,autodefrag 0 0
#
UUID=0e0e2f5d-a93f-4633-ba12-15a06763d641 /mnt/Disk_D/Games btrfs subvol=@Games,defaults,max_inline=256,noatime,space_cache=v2,autodefrag,compress=zstd:15 0 0
#
UUID=0e0e2f5d-a93f-4633-ba12-15a06763d641 /mnt/Disk_D/Igri btrfs subvol=@igri,defaults,max_inline=256,noatime,space_cache=v2,autodefrag,compress=zstd:15 0 0
#
UUID=0e0e2f5d-a93f-4633-ba12-15a06763d641 /mnt/Disk_D/Konsol btrfs subvol=@konsol,defaults,max_inline=256,noatime,space_cache=v2,autodefrag,compress=zstd:15 0 0
#
UUID=0e0e2f5d-a93f-4633-ba12-15a06763d641 /mnt/Disk_D/Torrent btrfs subvol=@torrent,defaults,nodatacow,max_inline=256,noatime,space_cache=v2 0 0
#
UUID=0e0e2f5d-a93f-4633-ba12-15a06763d641 /mnt/Disk_D/Swap btrfs subvol=@swap,defaults,nodatacow,noatime 0 0
```
The swap was located inside the file.

The only thing is, I did not periodically reset the cache of free space (I met an opinion on the Internet that this operation is needed, but I'm not sure.)
Well, accordingly, I have never performed balancing, since I have BtrFS on one physical medium.

As for the core, I don't remember. The file system is already 1.5-2 years old. Since I have an Arch-based distro, I use the newest kernel possible.

**Comment by Vochatrak-az-ezm** [comment]
After **btrfs check --init-csum-tree** 
```
# btrfs check --force --readonly /dev/sdd2
Opening filesystem to check...
Checking filesystem on /dev/sdd2
UUID: 0e0e2f5d-a93f-4633-ba12-15a06763d641
[1/7] checking root items
[2/7] checking extents
[3/7] checking free space cache
[4/7] checking fs roots
root 258 inode 271 errors 800, odd csum item
root 258 inode 612 errors 800, odd csum item
root 258 inode 613 errors 800, odd csum item
root 258 inode 614 errors 800, odd csum item
root 258 inode 615 errors 800, odd csum item
root 258 inode 616 errors 800, odd csum item
root 258 inode 617 errors 800, odd csum item
root 258 inode 618 errors 800, odd csum item
root 258 inode 898 errors 800, odd csum item
root 258 inode 2194 errors 800, odd csum item
root 258 inode 2319 errors 800, odd csum item
root 258 inode 2320 errors 800, odd csum item
root 258 inode 2321 errors 800, odd csum item
root 258 inode 2333 errors 800, odd csum item
root 258 inode 2334 errors 800, odd csum item
root 258 inode 2335 errors 800, odd csum item
root 258 inode 2338 errors 800, odd csum item
root 258 inode 2339 errors 800, odd csum item
root 258 inode 2340 errors 800, odd csum item
root 258 inode 2341 errors 800, odd csum item
root 258 inode 2342 errors 800, odd csum item
root 258 inode 3493 errors 800, odd csum item
root 258 inode 3494 errors 800, odd csum item
root 258 inode 3513 errors 800, odd csum item
root 258 inode 3516 errors 800, odd csum item
root 258 inode 3519 errors 800, odd csum item
root 258 inode 3522 errors 800, odd csum item
root 258 inode 3523 errors 800, odd csum item
root 258 inode 3524 errors 800, odd csum item
root 258 inode 3527 errors 800, odd csum item
root 258 inode 3528 errors 800, odd csum item
root 258 inode 3531 errors 800, odd csum item
root 258 inode 3532 errors 800, odd csum item
root 258 inode 3534 errors 800, odd csum item
root 258 inode 3535 errors 800, odd csum item
root 258 inode 3536 errors 800, odd csum item
root 258 inode 3538 errors 800, odd csum item
root 258 inode 3539 errors 800, odd csum item
root 258 inode 3540 errors 800, odd csum item
root 258 inode 3541 errors 800, odd csum item
root 258 inode 3542 errors 800, odd csum item
root 258 inode 3543 errors 800, odd csum item
root 258 inode 3545 errors 800, odd csum item
root 258 inode 3547 errors 800, odd csum item
root 258 inode 3548 errors 800, odd csum item
root 258 inode 3549 errors 800, odd csum item
root 258 inode 3613 errors 800, odd csum item
root 258 inode 3614 errors 800, odd csum item
root 258 inode 3615 errors 800, odd csum item
root 258 inode 3617 errors 800, odd csum item
root 258 inode 3618 errors 800, odd csum item
root 258 inode 3619 errors 800, odd csum item
root 258 inode 3630 errors 800, odd csum item
root 258 inode 3632 errors 800, odd csum item
root 258 inode 3634 errors 800, odd csum item
root 258 inode 3662 errors 800, odd csum item
root 258 inode 3674 errors 800, odd csum item
root 258 inode 3736 errors 800, odd csum item
root 258 inode 3746 errors 800, odd csum item
root 258 inode 3990 errors 800, odd csum item
root 258 inode 3991 errors 800, odd csum item
root 258 inode 3992 errors 800, odd csum item
root 258 inode 3993 errors 800, odd csum item
root 258 inode 3994 errors 800, odd csum item
root 258 inode 3995 errors 800, odd csum item
root 258 inode 3996 errors 800, odd csum item
root 258 inode 3997 errors 800, odd csum item
root 258 inode 4020 errors 800, odd csum item
root 258 inode 4021 errors 800, odd csum item
root 258 inode 4022 errors 800, odd csum item
root 258 inode 4023 errors 800, odd csum item
root 258 inode 4024 errors 800, odd csum item
root 258 inode 4036 errors 800, odd csum item
root 258 inode 4037 errors 800, odd csum item
root 258 inode 4047 errors 800, odd csum item
root 258 inode 4175 errors 800, odd csum item
root 258 inode 4193 errors 800, odd csum item
root 258 inode 4194 errors 800, odd csum item
root 258 inode 4196 errors 800, odd csum item
root 258 inode 4199 errors 800, odd csum item
root 258 inode 4200 errors 800, odd csum item
root 258 inode 4278 errors 800, odd csum item
root 258 inode 4281 errors 800, odd csum item
root 258 inode 4282 errors 800, odd csum item
root 258 inode 4283 errors 800, odd csum item
root 258 inode 4284 errors 800, odd csum item
root 258 inode 4286 errors 800, odd csum item
root 258 inode 4290 errors 800, odd csum item
root 258 inode 4293 errors 800, odd csum item
root 258 inode 4294 errors 800, odd csum item
root 258 inode 4295 errors 800, odd csum item
root 258 inode 4296 errors 800, odd csum item
root 258 inode 4300 errors 800, odd csum item
root 258 inode 4313 errors 800, odd csum item
root 258 inode 4316 errors 800, odd csum item
root 258 inode 4317 errors 800, odd csum item
root 258 inode 4321 errors 800, odd csum item
root 258 inode 4322 errors 800, odd csum item
root 258 inode 4323 errors 800, odd csum item
root 258 inode 4324 errors 800, odd csum item
root 258 inode 4325 errors 800, odd csum item
root 258 inode 4326 errors 800, odd csum item
root 258 inode 4327 errors 800, odd csum item
root 258 inode 4328 errors 800, odd csum item
root 258 inode 4329 errors 800, odd csum item
root 258 inode 4333 errors 800, odd csum item
root 258 inode 4354 errors 800, odd csum item
root 258 inode 4355 errors 800, odd csum item
root 258 inode 4356 errors 800, odd csum item
root 258 inode 4357 errors 800, odd csum item
root 258 inode 4358 errors 800, odd csum item
root 258 inode 4359 errors 800, odd csum item
root 258 inode 4360 errors 800, odd csum item
root 258 inode 4361 errors 800, odd csum item
root 258 inode 4362 errors 800, odd csum item
root 258 inode 4363 errors 800, odd csum item
root 258 inode 4364 errors 800, odd csum item
root 258 inode 4365 errors 800, odd csum item
root 258 inode 4366 errors 800, odd csum item
root 258 inode 4367 errors 800, odd csum item
root 258 inode 4368 errors 800, odd csum item
root 258 inode 4369 errors 800, odd csum item
root 258 inode 4370 errors 800, odd csum item
root 258 inode 4371 errors 800, odd csum item
root 258 inode 4372 errors 800, odd csum item
root 258 inode 4373 errors 800, odd csum item
root 258 inode 4374 errors 800, odd csum item
root 258 inode 4375 errors 800, odd csum item
root 258 inode 4376 errors 800, odd csum item
root 258 inode 4378 errors 800, odd csum item
root 258 inode 4379 errors 800, odd csum item
root 258 inode 4380 errors 800, odd csum item
root 258 inode 4381 errors 800, odd csum item
root 258 inode 4382 errors 800, odd csum item
root 258 inode 4383 errors 800, odd csum item
root 258 inode 4384 errors 800, odd csum item
root 258 inode 4385 errors 800, odd csum item
root 258 inode 4386 errors 800, odd csum item
root 258 inode 4387 errors 800, odd csum item
root 258 inode 4388 errors 800, odd csum item
root 258 inode 4389 errors 800, odd csum item
root 258 inode 4390 errors 800, odd csum item
root 258 inode 4391 errors 800, odd csum item
root 258 inode 4392 errors 800, odd csum item
root 258 inode 4393 errors 800, odd csum item
root 258 inode 4394 errors 800, odd csum item
root 258 inode 4395 errors 800, odd csum item
root 258 inode 4396 errors 800, odd csum item
root 258 inode 4397 errors 800, odd csum item
root 258 inode 4398 errors 800, odd csum item
root 258 inode 4399 errors 800, odd csum item
root 258 inode 4400 errors 800, odd csum item
root 258 inode 4401 errors 800, odd csum item
root 258 inode 4402 errors 800, odd csum item
root 258 inode 4403 errors 800, odd csum item
root 258 inode 4404 errors 800, odd csum item
root 258 inode 4405 errors 800, odd csum item
root 258 inode 4406 errors 800, odd csum item
root 258 inode 4407 errors 800, odd csum item
root 258 inode 4408 errors 800, odd csum item
root 258 inode 4409 errors 800, odd csum item
root 258 inode 4410 errors 800, odd csum item
root 258 inode 4411 errors 800, odd csum item
root 258 inode 4412 errors 800, odd csum item
root 258 inode 4413 errors 800, odd csum item
root 258 inode 4414 errors 800, odd csum item
root 258 inode 4416 errors 800, odd csum item
root 258 inode 4417 errors 800, odd csum item
root 258 inode 4419 errors 800, odd csum item
root 258 inode 4420 errors 800, odd csum item
root 258 inode 4423 errors 800, odd csum item
root 258 inode 4424 errors 800, odd csum item
root 258 inode 4425 errors 800, odd csum item
root 258 inode 4539 errors 800, odd csum item
root 258 inode 4540 errors 800, odd csum item
root 258 inode 4579 errors 800, odd csum item
root 258 inode 4853 errors 800, odd csum item
root 258 inode 4855 errors 800, odd csum item
root 258 inode 4856 errors 800, odd csum item
root 258 inode 4857 errors 800, odd csum item
root 258 inode 4858 errors 800, odd csum item
root 258 inode 4859 errors 800, odd csum item
root 258 inode 4860 errors 800, odd csum item
root 258 inode 4861 errors 800, odd csum item
root 258 inode 4862 errors 800, odd csum item
root 258 inode 4863 errors 800, odd csum item
root 258 inode 4864 errors 800, odd csum item
root 258 inode 4865 errors 800, odd csum item
root 258 inode 4891 errors 800, odd csum item
root 258 inode 4892 errors 800, odd csum item
root 258 inode 4909 errors 800, odd csum item
root 258 inode 4910 errors 800, odd csum item
root 258 inode 4935 errors 800, odd csum item
root 258 inode 4936 errors 800, odd csum item
root 258 inode 4937 errors 800, odd csum item
root 258 inode 4939 errors 800, odd csum item
root 258 inode 4940 errors 800, odd csum item
root 258 inode 4941 errors 800, odd csum item
root 258 inode 4952 errors 800, odd csum item
root 258 inode 4958 errors 800, odd csum item
root 258 inode 4979 errors 800, odd csum item
root 258 inode 6928 errors 800, odd csum item
root 258 inode 11105 errors 800, odd csum item
root 258 inode 12773 errors 800, odd csum item
root 258 inode 12774 errors 800, odd csum item
root 258 inode 12775 errors 800, odd csum item
root 258 inode 12776 errors 800, odd csum item
root 258 inode 12853 errors 800, odd csum item
root 258 inode 12854 errors 800, odd csum item
root 258 inode 12855 errors 800, odd csum item
root 258 inode 12856 errors 800, odd csum item
root 258 inode 12857 errors 800, odd csum item
root 258 inode 12860 errors 800, odd csum item
root 258 inode 12861 errors 800, odd csum item
root 258 inode 12862 errors 800, odd csum item
root 258 inode 12863 errors 800, odd csum item
root 258 inode 12864 errors 800, odd csum item
root 258 inode 12865 errors 800, odd csum item
root 258 inode 12866 errors 800, odd csum item
root 258 inode 12867 errors 800, odd csum item
root 258 inode 12868 errors 800, odd csum item
root 258 inode 12869 errors 800, odd csum item
root 258 inode 12870 errors 800, odd csum item
root 258 inode 12871 errors 800, odd csum item
root 258 inode 12872 errors 800, odd csum item
root 258 inode 12880 errors 800, odd csum item
root 258 inode 12924 errors 800, odd csum item
root 258 inode 12925 errors 800, odd csum item
root 258 inode 12926 errors 800, odd csum item
root 258 inode 12927 errors 800, odd csum item
root 258 inode 12928 errors 800, odd csum item
root 258 inode 12929 errors 800, odd csum item
root 258 inode 12931 errors 800, odd csum item
root 258 inode 12932 errors 800, odd csum item
root 258 inode 12933 errors 800, odd csum item
root 258 inode 12934 errors 800, odd csum item
root 258 inode 12935 errors 800, odd csum item
root 258 inode 12939 errors 800, odd csum item
root 258 inode 12940 errors 800, odd csum item
root 258 inode 12945 errors 800, odd csum item
root 258 inode 12952 errors 800, odd csum item
root 258 inode 12953 errors 800, odd csum item
root 258 inode 12955 errors 800, odd csum item
root 258 inode 12956 errors 800, odd csum item
root 258 inode 12957 errors 800, odd csum item
root 258 inode 12958 errors 800, odd csum item
root 258 inode 12959 errors 800, odd csum item
root 258 inode 13154 errors 800, odd csum item
root 258 inode 13155 errors 800, odd csum item
root 258 inode 13156 errors 800, odd csum item
root 258 inode 13220 errors 800, odd csum item
root 258 inode 13229 errors 800, odd csum item
root 258 inode 13230 errors 800, odd csum item
root 258 inode 13257 errors 800, odd csum item
root 258 inode 13259 errors 800, odd csum item
root 258 inode 13260 errors 800, odd csum item
root 258 inode 13261 errors 800, odd csum item
root 258 inode 13262 errors 800, odd csum item
root 258 inode 13267 errors 800, odd csum item
root 258 inode 13269 errors 800, odd csum item
root 258 inode 13270 errors 800, odd csum item
root 258 inode 13271 errors 800, odd csum item
root 258 inode 13273 errors 800, odd csum item
root 258 inode 13275 errors 800, odd csum item
root 258 inode 13276 errors 800, odd csum item
root 258 inode 13395 errors 800, odd csum item
root 258 inode 13396 errors 800, odd csum item
ERROR: errors found in fs roots
found 685525491712 bytes used, error(s) found
total csum bytes: 668696776
total tree bytes: 779206656
total fs tree bytes: 34062336
total extent tree bytes: 27426816
btree space waste bytes: 55859487
file data blocks allocated: 688818065408
 referenced 690986553344
```

**Comment by adam900710** [comment]
`Odd csum item` means pre-allocated extent has some csum.

After checking the code, the csum initialization code doesn't take pre-allocated extent into consideration.

Just to be extra safe, can you run `btrfs check --readonly --mode=lowmem` to provide a more specific report?

Thanks.

**Comment by adam900710** [comment]
After more digging into the kernel and some experiments, yes, this is indeed a bug in `--init-csum-tree`.

And it will further cause some false btrfs check (original mode bug) alerts in the future. But kernel should be able to handle such situation without much hassle.

So there is nothing to worry about for now, just wasted some metadata space for those pre-allocated space.

And if you filled those preallocated space in the future, the problem should be gone naturely.

**Comment by Vochatrak-az-ezm** [comment]
**btrfs check --readonly --mode=lowmem**
It gives a huge number of errors like ```ERROR: root 258 EXTENT_DATA[13396 2945409024] nodatasum shouldn't have datasum```

Do I understand correctly that defragmenting or balancing the partition should completely eliminate the problem?

**Comment by adam900710** [comment]
Defrag, at least `btrfs filesystem defrag` won't touch preallocated space.

Balance won't change pre-allocated space either.

Thus none of the above operations would fix the minor problem.

I'll work to fix the csum tree init code soon, but for now, you don't need to worry much, and can use the fs as usual.

**Comment by Vochatrak-az-ezm** [comment]
Thanks.

**Comment by kdave** [comment]
Fixes to check --init-csum-tree released in 5.16.1.
